hs-bindgen/src/HsBindgen/C/Tc/Macro.hs

```diff
@@ -0,0 +1,1121 @@
+  testEquality FractionalTyCon FractionalTyCon = Just Refl
+  testEquality BitsTyCon       BitsTyCon       = Just Refl
+  testEquality _ _ = Nothing
+
+instance PrettyVal ( ClassTyCon n ) where
+  prettyVal = \case
+    EqTyCon         -> Pretty.Con "EqTyCon"         []
+    OrdTyCon        -> Pretty.Con "OrdTyCon"        []
+    NumTyCon        -> Pretty.Con "NumTyCon"        []
+    IntegralTyCon   -> Pretty.Con "IntegralTyCon"   []
+    FractionalTyCon -> Pretty.Con "FractionalTyCon" []
+    BitsTyCon       -> Pretty.Con "BitsTyCon"       []
+
+instance Show (TyCon n ki) where
+  show = \case
+    DataTyCon  tc -> show tc
+    ClassTyCon tc -> show tc
+
+instance Show (DataTyCon n) where
+  show = \case
+    BoolTyCon   -> "Bool"
+    StringTyCon -> "String"
+    IntTyCon    -> "Int"
+    DoubleTyCon -> "Double"
+instance Show (ClassTyCon n) where
+  show = \case
+    EqTyCon         -> "Eq"
+    OrdTyCon        -> "Ord"
+    NumTyCon        -> "Num"
+    IntegralTyCon   -> "Integral"
+    FractionalTyCon -> "Fractional"
+    BitsTyCon       -> "Bits"
+
+-- | On-the-nose type equality.
+eqType :: Type Ty -> Type Ty -> Bool
+eqType ( TyVarTy tv1 ) ( TyVarTy tv2 ) = tyVarUnique tv1 == tyVarUnique tv2
+eqType ( TyConAppTy ( DataTyCon tc1 ) args1 ) ( TyConAppTy ( DataTyCon tc2 ) args2 )
+  = case testEquality tc1 tc2 of
+      Nothing -> False
+      Just Refl ->
+        eqTypes ( Vec.zipWith (,) args1 args2 )
+eqType ( FunTy args1 res1 ) ( FunTy args2 res2 )
+  =  length args1 == length args2
+  && eqTypes ( NE.zip args1 args2 )
+  && eqType res1 res2
+eqType _ _ = False
+
+eqTypes :: ( Functor f, Foldable f ) => f ( Type Ty, Type Ty ) -> Bool
+eqTypes = and . fmap ( uncurry eqType )
+
+{-------------------------------------------------------------------------------
+  Free type variables and substitution
+-------------------------------------------------------------------------------}
+
+freeTyVarsOfType :: forall ki. HasCallStack => IntSet -> Type ki -> [ TyVar ]
+freeTyVarsOfType bound0 = nubBy ( (==) `on` tyVarUnique ) . reverse . goTy bound0 []
+  where
+    goTy :: forall ki'. IntSet -> [ TyVar ] -> Type ki' -> [ TyVar ]
+    goTy bound !tvs = \case
+      TyVarTy tv ->
+        let u = uniqueInt $ tyVarUnique tv
+        in if u `IntSet.member` bound
+           then tvs
+           else tv : tvs
+      FunTy args res -> goFunTy bound tvs args res
+      TyConAppTy _tc tys -> concatMap (freeTyVarsOfType bound) tys
+
+    goFunTy :: forall ki'. IntSet -> [ TyVar ] -> NE.NonEmpty ( Type Ty ) -> Type ki' -> [ TyVar ]
+    goFunTy bound tvs (argTy NE.:| mbArgTys) resTy =
+      let !tvs' = goTy bound tvs argTy
+      in case NE.nonEmpty mbArgTys of
+        Nothing     -> goTy bound tvs resTy
+        Just argTys -> goFunTy bound tvs' argTys resTy
+
+tyVarSet :: [ TyVar ] -> IntSet
+tyVarSet = IntSet.fromList . map ( uniqueInt . tyVarUnique )
+
+newtype Subst tv = Subst { substMap :: IntMap ( tv, Type Ty ) }
+  deriving newtype Monoid
```

Comment:
this makes `mappend` not equal to your `<>`.

Reply:
Yes indeed, thanks. Fixed.